### PR TITLE
feat/test(CL Fees): Swap fees for in given out

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #4088

## What is the purpose of the change

This PR implements swap fees for swaps in given out. Refactors the `sympy` estimation scripts.

Refactors and improves the Go swap test suite.

## Brief Changelog

- reactors `sympy` scripts to estimate the results for the new tests
- implements the fee logic
- adds the swap fee tests from `sympy` estimates
- small clean up in `swaps_test.go`
- run `go mod tidy` and `go work sync`

## Testing and Verifying

This change added tests

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

**Comment by p0mvn** [comment]
This is ready but blocked on https://github.com/osmosis-labs/osmosis/pull/4127 being merged

File: x/concentrated-liquidity/fees.go
Comment (mattverse) [inline x/concentrated-liquidity/fees.go:290]: just curious, what led us to the decision to round this up?

File: x/concentrated-liquidity/internal/math/math.go
Comment (mattverse) [inline x/concentrated-liquidity/internal/math/math.go:151]: what's the signifiance of adding smallest dec?

File: x/concentrated-liquidity/swaps.go
Comment (mattverse) [inline x/concentrated-liquidity/swaps.go:478]: AFAIR, doesnt this function return amountIn and out, respectively?(implying that the order we used before was correct?)

File: x/concentrated-liquidity/swaps.go
Comment (mattverse) [inline x/concentrated-liquidity/swaps.go:492]: What's the reason we don't subtract the value that has been applied swap fee?

File: x/concentrated-liquidity/swaps_test.go
Comment (mattverse) [inline x/concentrated-liquidity/swaps_test.go:808]: love this format, nice!

**Review by mattverse — COMMENTED** [review]
Mostly questions on why things happened / or did not happen. Otherwise lgtm

File: scripts/cl/one_for_zero.py
Comment (czarcas7ic) [inline scripts/cl/one_for_zero.py:46]: why do we assign `token_in` to a func named `get_token_out`?

File: scripts/cl/one_for_zero.py
Comment (czarcas7ic) [inline scripts/cl/one_for_zero.py:86]: same question here, maybe we just need to rename the func?

File: x/concentrated-liquidity/swaps_test.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/swaps_test.go:865]: (.5% fee)

File: x/concentrated-liquidity/swaps_test.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/swaps_test.go:880]: (3% fee)

File: x/concentrated-liquidity/swaps_test.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/swaps_test.go:895]: (1% fee)

**Review by czarcas7ic — COMMENTED** [review]
Really well done on the python script as well as cleaning up the test cases! I have the same questions as Matt did with regards to the amountOut/In changes in swaps.go. Happy to approve once that is clarified.

File: x/concentrated-liquidity/fees.go
Comment (p0mvn) [reply x/concentrated-liquidity/fees.go:290]: This is pretty minor but just a safety precaution.

Default `Mul` does banker's rounding. However, we want to always round up the fee so that we don't undercharge it

File: x/concentrated-liquidity/swaps.go
Comment (p0mvn) [reply x/concentrated-liquidity/swaps.go:492]: The fee is always charged on the amount of token in

File: x/concentrated-liquidity/swaps.go
Comment (p0mvn) [reply x/concentrated-liquidity/swaps.go:478]: In the context of swap out given in you would be right.

However, here everything is reversed since now we are giving amount out as an input to this function.

We need to refactor `ComputeSwapStep` to be more generalized with the names of the internal variables and its returns.

I left it out of this PR as it is out of scope. I will add this as TODO for https://github.com/osmosis-labs/osmosis/issues/3218

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [reply x/concentrated-liquidity/internal/math/math.go:151]: By first truncating and then adding the smallest value, we turn this operation into multiplication that rounds up

File: scripts/cl/one_for_zero.py
Comment (p0mvn) [reply scripts/cl/one_for_zero.py:46]: Similar idea as in this discussion with the Go tests:
https://github.com/osmosis-labs/osmosis/pull/4131#discussion_r1091867438

I can make a wrapper function `get_token_in_swap_in_given_out` around `get_token_out` and call `get_token_in_swap_in_given_out` here to make it clearer

File: scripts/cl/one_for_zero.py
Comment (p0mvn) [reply scripts/cl/one_for_zero.py:86]: If we rename, it would break the logic for swaps out given in.

Suggesting this approach instead:
https://github.com/osmosis-labs/osmosis/pull/4131#discussion_r1092796608

**Comment by p0mvn** [comment]
Got back to all comments, going to make all the requested changes before merge

**Comment by mattverse** [comment]
@p0mvn Can you re-assign me as reviewer once r4r?

File: scripts/cl/one_for_zero.py
Comment (p0mvn) [reply scripts/cl/one_for_zero.py]: Done: [c3d6116](https://github.com/osmosis-labs/osmosis/pull/4131/commits/c3d61168b29cee677edca8ab38cd6b783c08a409)

**Review by czarcas7ic — APPROVED** [review]
Approving as I am satisfied with the responses to Matt's questions 👍

File: x/concentrated-liquidity/fees.go
Comment (mattverse) [reply x/concentrated-liquidity/fees.go:290]: makes sense, can we have in line doc regarding this on this line on why we mul round up here? I think it would be useful context for future code mangement

**Review by mattverse — APPROVED** [review]
LGTM for me as well once the request for my in line comment has been addressed 🌮

**Pull request merged**
